### PR TITLE
feat(ui): withTeamsForUser HOC conditionally fetch teams

### PR DESCRIPTION
**Pull request by davidenwang**
Before deciding to fetch teams for a user, the `withTeamsForUser` HOC will now check if there is information within the organization that it can use to pass down teams (hydrated with project information) to its wrapped components. I also fixed a small inconsistency with this component as it should not be setting the project store because it only contains a user's projects not all the projects they have access to.

### CODE REVIEW

File: src/sentry/static/sentry/app/utils/withTeamsForUser.tsx [hunk 1]
```diff
@@ -39,20 +40,34 @@ const withTeamsForUser = <P extends InjectedTeamsProps>(
...
 
     async fetchTeams() {
+      // check if we can use organization teams/projects instead of fetching data
+      const {projects} = this.props.organization;
+      let {teams} = this.props.organization;
+      if (projects && teams) {
+        const {isSuperuser} = ConfigStore.get('user');
+        const {projectsByTeam} = getProjectsByTeams(teams, projects, isSuperuser);
+        teams.forEach(team => {
+          team.projects = projectsByTeam[team.slug] || [];
```
Comment (scefali) [inline src/sentry/static/sentry/app/utils/withTeamsForUser.tsx]: @davidenwang This is going to mutate your props which could produce unintended side effects. We should make a copy (e.g. `Object.assign`) instead.

File: src/sentry/static/sentry/app/types/index.tsx [hunk 2]
```diff
@@ -56,6 +56,7 @@ export type Team = {
   id: string;
   slug: string;
   isMember: boolean;
+  projects?: Project[];
```
[the remark this answers is not in the document]
Comment (davidenwang) [reply src/sentry/static/sentry/app/types/index.tsx]: oooh yea that sounds good

File: src/sentry/static/sentry/app/utils/withTeamsForUser.tsx [hunk 3]
```diff
@@ -62,6 +72,18 @@ const withTeamsForUser = <P extends InjectedTeamsProps>(
       }
     }
 
+    populateTeamsWithProjects(teams: Team[], projects: Project[], isSuperuser: boolean) {
```
Comment (scefali) [inline src/sentry/static/sentry/app/utils/withTeamsForUser.tsx]: @davidenwang You might consider omitting the `isSuperuser` input and just reading it from `ConfigStore`. Just a thought.

**Review by scefali — APPROVED** [review]

File: src/sentry/static/sentry/app/utils/withTeamsForUser.tsx (hunk 3 above)
Comment (davidenwang) [reply src/sentry/static/sentry/app/utils/withTeamsForUser.tsx]: mm yea that makes more sense

**Review by billyvg — APPROVED** [review]

**Pull request merged**